panic when starting multiple servers concurrently

We at [Ably](https://www.ably.io) have been testing our Liftbridge integration and have been experiencing the following panic when starting multiple servers concurrently:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1a0dc09]

goroutine 331 [running]:
github.com/liftbridge-io/liftbridge/server.(*activityManager).createActivityStream(0xc0005be640, 0x0, 0xc0006f6028)
        /Users/lewis/go/pkg/mod/github.com/liftbridge-io/liftbridge@v1.3.0/server/activity.go:210 +0x179
```

Looking at the code, this panic suggests that the server's loopback client is not yet created when it's promoted to a leader, and it tries to use a `nil` client to create an activity stream [here](https://github.com/liftbridge-io/liftbridge/blob/v1.3.0/server/activity.go#L210).

I was able to reproduce this using the following test case (although the panic is intermittent):

```go
// Ensure servers with activity stream enabled can be bootstrapped concurrently
func TestBootstrapConcurrentWithActivityStream(t *testing.T) {
	defer cleanupStorage(t)

	// Use a central NATS server.
	ns := natsdTest.RunDefaultServer()
	defer ns.Shutdown()

	// Start three servers concurrently with activity stream enabled
	ids := []string{"s1", "s2", "s3"}
	type result struct {
		server *Server
		err    error
	}
	results := make(chan *result, 3)
	done := make(chan struct{})
	defer close(done)
	for _, id := range ids {
		go func(id string) {
			config := getTestConfig(id, false, 0)
			config.Clustering.ServerID = id
			config.Clustering.RaftBootstrapPeers = ids
			config.ActivityStream.Enabled = true

			server, err := RunServerWithConfig(config)

			select {
			case results <- &result{server, err}:
			case <-done:
			}
		}(id)
	}

	// wait for the servers to start
	for i := 0; i < 3; i++ {
		res := <-results
		if res.err != nil {
			t.Error(res.err)
		} else {
			defer res.server.Stop()
		}
	}
}
```

Here's a good test run:

```
$ go test -v --run TestBootstrapConcurrentWithActivityStream -count=1 ./server
=== RUN   TestBootstrapConcurrentWithActivityStream
--- PASS: TestBootstrapConcurrentWithActivityStream (5.26s)
PASS
ok      github.com/liftbridge-io/liftbridge/server      5.287s
```

and here's a bad one:

```
$ go test -v --run TestBootstrapConcurrentWithActivityStream -count=1 ./server
=== RUN   TestBootstrapConcurrentWithActivityStream
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1982389]

goroutine 153 [running]:
github.com/liftbridge-io/liftbridge/server.(*activityManager).createActivityStream(0xc000130ec0, 0x0, 0xc0001685a8)
        /Users/lewis/src/github.com/liftbridge-io/liftbridge/server/activity.go:210 +0x179
github.com/liftbridge-io/liftbridge/server.(*activityManager).BecomeLeader(0xc000130ec0, 0xc000122240, 0x2a)
        /Users/lewis/src/github.com/liftbridge-io/liftbridge/server/activity.go:72 +0x44
github.com/liftbridge-io/liftbridge/server.(*Server).leadershipAcquired(0xc0003b4000, 0xc0004c3f08, 0x2)
        /Users/lewis/src/github.com/liftbridge-io/liftbridge/server/server.go:595 +0x1e4
github.com/liftbridge-io/liftbridge/server.(*Server).startMetadataRaft.func1()
        /Users/lewis/src/github.com/liftbridge-io/liftbridge/server/server.go:533 +0x1ba
github.com/liftbridge-io/liftbridge/server.(*Server).startGoroutine.func1(0xc0001a1640, 0xc0003b4000)
        /Users/lewis/src/github.com/liftbridge-io/liftbridge/server/server.go:958 +0x27
created by github.com/liftbridge-io/liftbridge/server.(*Server).startGoroutine
        /Users/lewis/src/github.com/liftbridge-io/liftbridge/server/server.go:957 +0x97
FAIL    github.com/liftbridge-io/liftbridge/server      1.505s
FAIL
```

I _think_ the issue is that as part of initialising a loopback client, the server broadcasts a message to gather info from the other peers and waits for them all to respond [here](https://github.com/liftbridge-io/liftbridge/blob/v1.3.0/server/metadata.go#L231-L256), which for `s1` likely blocks as the other servers were not yet listening for that request, but then if `s1` gets promoted to a leader, it performs its leader duties even though its loopback client request is still blocked.

Perhaps there needs to be synchronisation between setting the loopback client and creating the activity stream.

--- a/server/activity.go
+++ b/server/activity.go
@@ -84,7 +84,9 @@ func (a *activityManager) BecomeFollower() error {
 		return nil
 	}
 
-	close(a.leadershipLostCh)
+	if a.leadershipLostCh != nil {
+		close(a.leadershipLostCh)
+	}
 	return nil
 }
 

--- a/server/activity.go
+++ b/server/activity.go
@@ -2,14 +2,14 @@ package server
 
 import (
 	"context"
-	"fmt"
 	"sync"
 	"time"
 
 	"github.com/hashicorp/raft"
-	lift "github.com/liftbridge-io/go-liftbridge/v2"
 	client "github.com/liftbridge-io/liftbridge-api/go"
 	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 
 	proto "github.com/liftbridge-io/liftbridge/server/protocol"
 )
@@ -209,12 +209,12 @@ func (a *activityManager) handleRaftLog(l *raft.Log) error {
 // createActivityStream creates the activity stream and connects a local client
 // that will be subscribed to it.
 func (a *activityManager) createActivityStream() error {
-	err := a.getLoopbackClient().CreateStream(context.Background(),
-		a.getActivityStreamSubject(),
-		activityStream,
-		lift.MaxReplication(),
-	)
-	if err != nil && err != lift.ErrStreamExists {
+	_, err := a.api.CreateStream(context.Background(), &client.CreateStreamRequest{
+		Subject:           a.getActivityStreamSubject(),
+		Name:              activityStream,
+		ReplicationFactor: -1,
+	})
+	if err != nil && status.Convert(err).Code() != codes.AlreadyExists {
 		return errors.Wrap(err, "failed to create an activity stream")
 	}
 
@@ -231,25 +231,11 @@ func (a *activityManager) publishActivityEvent(event *client.ActivityStreamEvent
 	ctx, cancel := context.WithTimeout(context.Background(), a.config.ActivityStream.PublishTimeout)
 	defer cancel()
 
-	var messageOption lift.MessageOption
-	ackPolicy := a.config.ActivityStream.PublishAckPolicy
-	switch ackPolicy {
-	case client.AckPolicy_LEADER:
-		messageOption = lift.AckPolicyLeader()
-	case client.AckPolicy_ALL:
-		messageOption = lift.AckPolicyAll()
-	case client.AckPolicy_NONE:
-		messageOption = lift.AckPolicyNone()
-	default:
-		panic(fmt.Sprintf("Unknown ack policy: %v", ackPolicy))
-	}
-
-	_, err = a.getLoopbackClient().Publish(
-		ctx,
-		activityStream,
-		data,
-		messageOption,
-	)
+	_, err = a.api.Publish(ctx, &client.PublishRequest{
+		Value:     data,
+		Stream:    activityStream,
+		AckPolicy: a.config.ActivityStream.PublishAckPolicy,
+	})
 	if err != nil {
 		return errors.Wrap(err, "failed to publish event to stream")
 	}

--- a/server/api.go
+++ b/server/api.go
@@ -166,6 +166,39 @@ func (a *apiServer) SetStreamReadonly(ctx context.Context, req *client.SetStream
 // messages when it reaches the end of the partition. Use the request context
 // to close the subscription.
 func (a *apiServer) Subscribe(req *client.SubscribeRequest, out client.API_SubscribeServer) error {
+	msgC, errC, cancel, err := a.SubscribeInternal(out.Context(), req)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	// Send an empty message which signals the subscription was successfully
+	// created.
+	if err := out.Send(&client.Message{}); err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-out.Context().Done():
+			return nil
+		case m := <-msgC:
+			if err := out.Send(m); err != nil {
+				return err
+			}
+		case err := <-errC:
+			return err.Err()
+		}
+	}
+}
+
+// Subscribe creates an ephemeral subscription for the given stream partition.
+// It begins to receive messages starting at the given offset and waits for new
+// messages when it reaches the end of the partition. Use the request context
+// to close the subscription. This is a non-gRPC API for internal use.
+func (a *apiServer) SubscribeInternal(ctx context.Context, req *client.SubscribeRequest) (
+	<-chan *client.Message, <-chan *status.Status, func(), error) {
+
 	a.logger.Debugf("api: Subscribe [stream=%s, partition=%d, start=%s, offset=%d, timestamp=%d]",
 		req.Stream, req.Partition, req.StartPosition, req.StartOffset, req.StartTimestamp)
 
@@ -174,7 +207,7 @@ func (a *apiServer) Subscribe(req *client.SubscribeRequest, out client.API_Subsc
 		a.logger.Errorf("api: Failed to subscribe to partition "+
 			"[stream=%s, partition=%d]: no such partition",
 			req.Stream, req.Partition)
-		return status.Error(codes.NotFound, "No such partition")
+		return nil, nil, nil, status.Error(codes.NotFound, "No such partition")
 	}
 
 	leader, _ := partition.GetLeader()
@@ -183,40 +216,18 @@ func (a *apiServer) Subscribe(req *client.SubscribeRequest, out client.API_Subsc
 			a.logger.Info("api: Accepting subscription to partition %s: server not stream leader", partition)
 		} else {
 			a.logger.Errorf("api: Failed to subscribe to partition %s: server not stream leader", partition)
-			return status.Error(codes.FailedPrecondition, "Server not partition leader")
+			return nil, nil, nil, status.Error(codes.FailedPrecondition, "Server not partition leader")
 		}
 	}
 
 	cancel := make(chan struct{})
-	defer close(cancel)
-	ch, errCh, err := a.subscribe(out.Context(), partition, req, cancel)
+	ch, errCh, err := a.subscribe(ctx, partition, req, cancel)
 	if err != nil {
 		a.logger.Errorf("api: Failed to subscribe to partition %s: %v", partition, err.Err())
-		return err.Err()
+		return nil, nil, nil, err.Err()
 	}
 
-	// Send an empty message which signals the subscription was successfully
-	// created.
-	if err := out.Send(&client.Message{}); err != nil {
-		return err
-	}
-
-	// Update the active subscriber count.
-	partition.IncreaseSubscriberCount()
-	defer partition.DecreaseSubscriberCount()
-
-	for {
-		select {
-		case <-out.Context().Done():
-			return nil
-		case m := <-ch:
-			if err := out.Send(m); err != nil {
-				return err
-			}
-		case err := <-errCh:
-			return err.Err()
-		}
-	}
+	return ch, errCh, func() { close(cancel) }, nil
 }
 
 // FetchMetadata retrieves the latest cluster metadata, including stream broker
@@ -681,6 +692,10 @@ func (a *apiServer) subscribe(ctx context.Context, partition *partition,
 	}
 
 	a.startGoroutine(func() {
+		// Update the active subscriber count.
+		partition.IncreaseSubscriberCount()
+		defer partition.DecreaseSubscriberCount()
+
 		headersBuf := make([]byte, 28)
 		for {
 			// TODO: this could be more efficient.

--- a/server/cursors.go
+++ b/server/cursors.go
@@ -8,10 +8,10 @@ import (
 	"time"
 
 	"github.com/hashicorp/golang-lru"
-	lift "github.com/liftbridge-io/go-liftbridge/v2"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 
+	client "github.com/liftbridge-io/liftbridge-api/go"
 	proto "github.com/liftbridge-io/liftbridge/server/protocol"
 )
 
@@ -112,8 +112,13 @@ func (c *cursorManager) SetCursor(ctx context.Context, streamName, cursorID stri
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	_, err = c.getLoopbackClient().Publish(ctx, cursorsStream, serializedCursor,
-		lift.ToPartition(cursorsPartitionID), lift.Key(cursorKey), lift.AckPolicyAll())
+	_, err = c.api.Publish(ctx, &client.PublishRequest{
+		Key:       cursorKey,
+		Value:     serializedCursor,
+		Stream:    cursorsStream,
+		Partition: cursorsPartitionID,
+		AckPolicy: client.AckPolicy_ALL,
+	})
 	if err != nil {
 		return status.New(codes.Internal, err.Error())
 	}
@@ -200,49 +205,35 @@ func (c *cursorManager) getLatestCursorOffset(ctx context.Context, cursorKey []b
 	// TODO: This can likely be made more efficient.
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	var (
-		errC = make(chan error, 1)
-		msgC = make(chan *lift.Message, 1)
-	)
-	err := c.getLoopbackClient().Subscribe(ctx, cursorsStream, func(msg *lift.Message, err error) {
-		defer func() {
-			if msg != nil && msg.Offset() == hw {
-				close(msgC)
-			}
-		}()
-		if err != nil {
-			errC <- err
-			return
-		}
-		if !bytes.Equal(msg.Key(), cursorKey) {
-			return
-		}
-		msgC <- msg
-
-	}, lift.Partition(partitionID), lift.StartAtEarliestReceived(), lift.Resume())
+	msgC, errC, cancel, err := c.api.SubscribeInternal(ctx, &client.SubscribeRequest{
+		Stream:        cursorsStream,
+		Partition:     partitionID,
+		StartPosition: client.StartPosition_EARLIEST,
+		Resume:        true,
+	})
 	if err != nil {
 		return 0, err
 	}
-
+	defer cancel()
 	var (
 		latestOffset = int64(-1)
 		cursor       = new(proto.Cursor)
 	)
 	for {
 		select {
-		case msg, ok := <-msgC:
-			if !ok {
+		case msg := <-msgC:
+			if bytes.Equal(msg.Key, cursorKey) {
+				if err := cursor.Unmarshal(msg.Value); err != nil {
+					c.logger.Errorf("Invalid cursor message in cursors stream: %v", err)
+				} else if cursor.Offset > latestOffset {
+					latestOffset = cursor.Offset
+				}
+			}
+			if msg.Offset == hw {
 				return latestOffset, nil
 			}
-			if err := cursor.Unmarshal(msg.Value()); err != nil {
-				c.logger.Errorf("Invalid cursor message in cursors stream: %v", err)
-				continue
-			}
-			if cursor.Offset > latestOffset {
-				latestOffset = cursor.Offset
-			}
 		case err := <-errC:
-			return 0, err
+			return 0, err.Err()
 		case <-ctx.Done():
 			return 0, ctx.Err()
 		}

--- a/server/server.go
+++ b/server/server.go
@@ -17,7 +17,6 @@ import (
 	"time"
 
 	"github.com/hashicorp/raft"
-	lift "github.com/liftbridge-io/go-liftbridge/v2"
 	client "github.com/liftbridge-io/liftbridge-api/go"
 	"github.com/nats-io/nats.go"
 	"github.com/nats-io/nuid"
@@ -55,7 +54,8 @@ type Server struct {
 	ncPublishes        *nats.Conn
 	logger             logger.Logger
 	loggerOut          io.Writer
-	api                *grpc.Server
+	grpcServer         *grpc.Server
+	api                *apiServer
 	metadata           *metadataAPI
 	shutdownCh         chan struct{}
 	raft               atomic.Value
@@ -66,7 +66,6 @@ type Server struct {
 	shutdown           bool
 	running            bool
 	goroutineWait      sync.WaitGroup
-	client             atomic.Value
 	activity           *activityManager
 	cursors            *cursorManager
 }
@@ -187,10 +186,6 @@ func (s *Server) Start() (err error) {
 		return errors.Wrap(err, "failed to start API server")
 	}
 
-	if err := s.createLoopbackClient(); err != nil {
-		return errors.Wrap(err, "failed to create loopback client")
-	}
-
 	s.startRaftLeadershipLoop()
 	return nil
 }
@@ -208,8 +203,8 @@ func (s *Server) Stop() error {
 	s.logger.Info("Shutting down...")
 
 	close(s.shutdownCh)
-	if s.api != nil {
-		s.api.Stop()
+	if s.grpcServer != nil {
+		s.grpcServer.Stop()
 	}
 
 	if s.listener != nil {
@@ -225,13 +220,6 @@ func (s *Server) Stop() error {
 
 	if raft := s.getRaft(); raft != nil {
 		if err := raft.shutdown(); err != nil {
-			s.mu.Unlock()
-			return err
-		}
-	}
-
-	if client := s.getLoopbackClient(); client != nil {
-		if err := client.Close(); err != nil {
 			s.mu.Unlock()
 			return err
 		}
@@ -413,18 +401,19 @@ func (s *Server) startAPIServer() error {
 		opts = append(opts, grpc.Creds(creds))
 	}
 
-	api := grpc.NewServer(opts...)
-	s.api = api
-	client.RegisterAPIServer(api, &apiServer{s})
+	grpcServer := grpc.NewServer(opts...)
+	s.grpcServer = grpcServer
+	s.api = &apiServer{s}
+	client.RegisterAPIServer(grpcServer, s.api)
 
-	health.Register(api)
+	health.Register(grpcServer)
 
 	s.mu.Lock()
 	s.running = true
 	s.mu.Unlock()
 	s.startGoroutine(func() {
 		health.SetServing()
-		err := api.Serve(s.listener)
+		err := grpcServer.Serve(s.listener)
 		s.mu.Lock()
 		s.running = false
 		s.mu.Unlock()
@@ -439,51 +428,6 @@ func (s *Server) startAPIServer() error {
 	})
 
 	return nil
-}
-
-// createLoopbackClient creates a Liftbridge client connected to this server's
-// loopback address used for internal purposes.
-func (s *Server) createLoopbackClient() error {
-	var (
-		addr     = s.listener.Addr()
-		deadline = time.Now().Add(10 * time.Second)
-		conn     net.Conn
-		err      error
-	)
-	for time.Now().Before(deadline) {
-		conn, err = net.Dial("tcp", addr.String())
-		if err == nil {
-			break
-		}
-		time.Sleep(time.Millisecond)
-	}
-	if conn == nil {
-		return err
-	}
-	if err := conn.Close(); err != nil {
-		return err
-	}
-	opts := []lift.ClientOption{}
-	if s.config.TLSCert != "" {
-		// Skip TLS verification since we're just connecting to loopback.
-		tlsConfig := &tls.Config{InsecureSkipVerify: true}
-		opts = append(opts, lift.TLSConfig(tlsConfig))
-	}
-	client, err := lift.Connect([]string{addr.String()}, opts...)
-	if err != nil {
-		return err
-	}
-	s.client.Store(client)
-	return nil
-}
-
-// getLoopbackClient returns the Liftbridge client connected to this server.
-func (s *Server) getLoopbackClient() lift.Client {
-	client := s.client.Load()
-	if client == nil {
-		return nil
-	}
-	return client.(lift.Client)
 }
 
 // createNATSConn creates a new NATS connection with the given name.

--- a/server/server.go
+++ b/server/server.go
@@ -163,7 +163,7 @@ func (s *Server) Start() (err error) {
 		s.config.Streams.SegmentMaxAge = time.Second
 	}
 
-	if err := s.startMetadataRaft(); err != nil {
+	if err := s.setupMetadataRaft(); err != nil {
 		return errors.Wrap(err, "failed to start Raft node")
 	}
 
@@ -187,7 +187,12 @@ func (s *Server) Start() (err error) {
 		return errors.Wrap(err, "failed to start API server")
 	}
 
-	return errors.Wrap(s.createLoopbackClient(), "failed to create loopback client")
+	if err := s.createLoopbackClient(); err != nil {
+		return errors.Wrap(err, "failed to create loopback client")
+	}
+
+	s.startRaftLeadershipLoop()
+	return nil
 }
 
 // Stop will attempt to gracefully shut the Server down by signaling the stop
@@ -515,16 +520,10 @@ func (s *Server) createNATSConn(name string) (*nats.Conn, error) {
 	return opts.Connect()
 }
 
-// startMetadataRaft creates and starts an embedded Raft node to participate in
-// the metadata cluster. This will bootstrap using the configured server
-// settings and start a goroutine for automatically responding to Raft
-// leadership changes.
-func (s *Server) startMetadataRaft() error {
-	if err := s.setupMetadataRaft(); err != nil {
-		return err
-	}
+// startRaftLeadershipLoop start a goroutine for automatically responding to
+// Raft leadership changes.
+func (s *Server) startRaftLeadershipLoop() {
 	node := s.getRaft()
-
 	s.startGoroutine(func() {
 		for {
 			select {
@@ -558,7 +557,6 @@ func (s *Server) startMetadataRaft() error {
 			}
 		}
 	})
-	return nil
 }
 
 // setRaft sets the Raft node for the server. This should only be called once

--- a/server/server_test.go
+++ b/server/server_test.go
@@ -491,6 +491,51 @@ func TestBootstrapMisconfiguration(t *testing.T) {
 	}
 }
 
+// Ensure servers with activity stream enabled can be bootstrapped
+// concurrently.
+func TestBootstrapConcurrentWithActivityStream(t *testing.T) {
+	defer cleanupStorage(t)
+
+	// Use a central NATS server.
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	// Start three servers concurrently with activity stream enabled.
+	ids := []string{"s1", "s2", "s3"}
+	type result struct {
+		server *Server
+		err    error
+	}
+	results := make(chan *result, 3)
+	done := make(chan struct{})
+	defer close(done)
+	for _, id := range ids {
+		go func(id string) {
+			config := getTestConfig(id, false, 0)
+			config.Clustering.ServerID = id
+			config.Clustering.RaftBootstrapPeers = ids
+			config.ActivityStream.Enabled = true
+
+			server, err := RunServerWithConfig(config)
+
+			select {
+			case results <- &result{server, err}:
+			case <-done:
+			}
+		}(id)
+	}
+
+	// Wait for the servers to start.
+	for i := 0; i < 3; i++ {
+		res := <-results
+		if res.err != nil {
+			t.Error(res.err)
+		} else {
+			defer res.server.Stop()
+		}
+	}
+}
+
 // Ensure when the metadata leader fails, a new one is elected.
 func TestMetadataLeaderFailover(t *testing.T) {
 	defer cleanupStorage(t)

